### PR TITLE
[sqla] Fixing ORDER BY logic

### PR DESCRIPTION
**Pull request by john-bodley**
### CATEGORY

Choose one

- [x] Bug Fix
- [ ] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

When upgrading Presto from v0.188 to v0.224 we noticed an issue with the `ORDER  BY` logic due to ambiguity which had changed, i.e., per [here](https://support.treasuredata.com/hc/en-us/articles/360008414094-Presto-0-188-to-0-205-Migration):

> The general rules for ORDER BY behavior are now:
> 
> - ORDER BY is executed after projection
> - ORDER BY prefers an ordinal alias wherever possible
> - Input and output columns can be used in the ORDER BY clause, but if there is ambiguity, the output columns have higher precedence

Though it's up to the individual SQLAlchemy dialect to handle how it translates a SQLAlchemy statement to a string we need to ensure that the SQLAlchemy statement is structured correctly. i.e., the `ORDER BY` clause references the same objects as the `SELECT` clause where appropriate. 

Here's a basic toy example, 

```python
from sqlalchemy.sql import literal_column, select

a = literal_column("SUM(x)").label("total")
b = literal_column("SUM(x)").label("total")
```

#### Current 

Here we are using **different** SQLAlchemy `Column` objects for the `SELECT` clause and `ORDER BY` clause expressions, akin to the current logic, where the output label isn't being rendered in the `ORDER BY` clause:

```python
print(str(select([a]).order_by(b)))
>>> SELECT SUM(x) AS total ORDER BY SUM(x)

```

#### Proposed

Here we are using the **same** SQLAlchemy `Column` objects for the `SELECT` clause and `ORDER BY` clause expressions, where the output label is being rendered in the `ORDER BY` clause.

```python
print(str(select([a]).order_by(a)))
>>> SELECT SUM(x) AS total ORDER BY total
```

For the table visualization type we add the `SORT BY` to the list of metrics (this is probably a requirement for some dialects) as the `ORDER BY` clause is executed after the projection (`SELECT`) and thus the `ORDER BY` expressions need to be present in the `SELECT` clause.

This PR simply ensures that we reference the metric object in the `ORDER BY` clause if defined. Note there are probably other places during the SQL construction phase where we're probably not building the query in a cohesive manner. 

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

to: @betodealmeida @michellethomas @mistercrunch @villebro

### CODE REVIEW

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -146,12 +146,12 @@ class TableColumn(Model, BaseColumn):
 
     def get_sqla_col(self, label: Optional[str] = None) -> Column:
         label = label or self.column_name
-        if not self.expression:
+        if self.expression:
```
Comment (john-bodley) [inline superset/connectors/sqla/models.py]: This logic is unchanged. I just removed the `not` and switched the order.

File: superset/connectors/sqla/models.py [hunk 2]
```diff
@@ -841,13 +841,27 @@ def get_sqla_query(  # sqla
...
+
+            if isinstance(col, Column):
+                sqla_col = col
+            else:
+                if utils.is_adhoc_metric(col):
+                    sqla_col = self.adhoc_metric_to_sqla(col, cols)
+                elif col in cols:
+                    sqla_col = cols[col].get_sqla_col()
+
+                if sqla_col is not None and str(sqla_col) in metrics_exprs_by_name:
```
Comment (john-bodley) [inline superset/connectors/sqla/models.py]: Matching of the string representation of a SQLAlchemy column seems kludgy though I'm not certain of a better approach. I initially was going to create an `OrderedDict` of `metrics` to `metric_expr` however the `metric` can be a `dict` which isn't hashable.

**Comment by villebro** [comment]
I wonder if we could instead ensure that the `.get_sqla_col()` methods in `TableColumn` and `SqlMetric` only instantiate each SqlAlchemy object once, and return the first instantiation upon subsequent calls. I believe this could be achieved by adding a `_sqla_col: Optional[Column]` attribute to said classes and return that if defined. This would save us from having having to make the downstream code aware of this problem.

**Comment by john-bodley** [comment]
@villebro great idea. I’ll make the change.

**Comment by villebro** [comment]
@john-bodley yes, that would be even DRYer. In addition to considering key column+label properties, I think memoization will probably need to include `db_engine_spec.engine` in the key to ensure engine specific behaviour doesn't leak.

**Comment by john-bodley** [comment]
@villebro the `db_engine_spec` is bound to the table per `self.table.database.db_engine_spec` and thus I’m unsure whether leaking is an issue, i.e., the column cache will be at the table level.

**Comment by villebro** [comment]
@john-bodley That's true; if the database id is included in the key this won't be a problem. I think if we have database id, column type (column or expression), value (column name or expression) and label, I think we should be ok.

**Comment by john-bodley** [comment]
@villebro reading through the code I’m somewhat hesitant to add a class level corpus of columns to the `Table` class especially as this is scoped to the SQL query for a specific query context.

An alternative would be to create a local corpus for each query however this would need to be passed around everywhere. We could make the SQL query it’s own class which stores the corpus, but I’m also concerned that we may be going down a rabbit hole.

**Comment by villebro** [comment]
@john_bodley agreed, major rabbit hole risk here, gives me flashbacks of `sqla/models.py` work in late 2018. I think we should do some more thorough refactoring + unit testing this time if we do decide to fix this more permanently. So perhaps just go ahead and merge this and defer the elegant solution to a later date.

**Comment by john-bodley** [comment]
@villebro agreed. The architecture of the sqla model should probably be refactored at some stage. Also there’s few few unit tests for the entire SQL generation, which we should address if we refactor, given how crucial this logic is.

**Review by villebro — APPROVED** [review]

**Pull request merged**